types/types.go

```diff
@@ -23,6 +24,20 @@ type Repository interface {
 	Image(ref string) Image // ref == image name w/o registry part
 }
 
+// ImageSource is a service, possibly remote (= slow), to download components of a single image.
+type ImageSource interface {
+	GetManifest() (manifest []byte, unverifiedCanonicalDigest string, err error)
+	GetLayer(digest string) (io.ReadCloser, error)
+	GetSignatures() ([][]byte, error)
+}
+
+// ImageDestination is a service, possibly remote (= slow), to store components of a single image.
+type ImageDestination interface {
+	PutManifest([]byte) error
+	PutLayer(digest string, stream io.ReadCloser) error
+	PutSignatures(signatures [][]byte) error
+}
```

Comment:
(I realize `docker save` streaming wouldn’t work with this interface, because we as the receiver can't dictate the order of layers coming from the stream; something to worry about later.)